static/panes/tool-input-view.js

```diff
@@ -0,0 +1,223 @@
+    this.container.on('destroy', this.close, this);
+
+    this.container.layoutManager.on('initialised', function () {
+        // Once initialized, let everyone know what text we have.
+        this.maybeEmitChange();
+    }, this);
+    this.eventHub.on('initialised', this.maybeEmitChange, this);
+
+    this.editor.getModel().onDidChangeContent(_.bind(function () {
+        this.debouncedEmitChange();
+        this.updateState();
+    }, this));
+
+    this.cursorSelectionThrottledFunction =
+        _.throttle(_.bind(this.onDidChangeCursorSelection, this), 500);
+    this.editor.onDidChangeCursorSelection(_.bind(function (e) {
+        this.cursorSelectionThrottledFunction(e);
+    }, this));
+};
+
+ToolInputView.prototype.setTitle = function () {
+    var name = this._toolName + ' Input #' + this._compilerId;
+    this.container.setTitle(name);
+};
+
+ToolInputView.prototype.resize = function () {
+    var topBarHeight = this.topBar.outerHeight(true);
+    this.editor.layout({
+        width: this.domRoot.width(),
+        height: this.domRoot.height() - topBarHeight,
+    });
+};
+
+ToolInputView.prototype.updateState = function () {
+    this.container.setState(this.currentState());
+};
+
+ToolInputView.prototype.currentState = function () {
+    var state = {
+        toolId: this._toolId,
+        toolName: this._toolName,
+        editorId: this._editorId,
+        compilerId: this._compilerId,
+    };
+    this.fontScale.addState(state);
+    return state;
+};
+
+ToolInputView.prototype.close = function () {
+    this.eventHub.unsubscribe();
+    this.eventHub.emit('toolInputViewClosed', this._compilerId, this._editorId, this._toolId, this.getInput());
+    this.editor.dispose();
+};
+
+ToolInputView.prototype.onToolClose = function (compilerId, toolSettings) {
+    if (this._compilerId === compilerId && this._toolId === toolSettings.toolId) {
+        // We can't immediately close as an outer loop somewhere in GoldenLayout is iterating over
+        // the hierarchy. We can't modify while it's being iterated over.
+        this.close();
+        _.defer(function (self) {
+            self.container.close();
+        }, this);
+    }
+};
+
+ToolInputView.prototype.onToolInputViewCloseRequest = function (compilerId, editorId, toolId) {
+    if (this._compilerId === compilerId && this._editorId === editorId && this._toolId === toolId) {
+        this.close();
+        _.defer(function (self) {
+            self.container.close();
+        }, this);
+    }
+};
+
+ToolInputView.prototype.onCompilerClose = function (id) {
+    if (id === this._compilerId) {
+        // We can't immediately close as an outer loop somewhere in GoldenLayout is iterating over
+        // the hierarchy. We can't modify while it's being iterated over.
+        this.close();
+        _.defer(function (self) {
+            self.container.close();
+        }, this);
+    }
+};
+
+ToolInputView.prototype.onSettingsChange = function (newSettings) {
```

Comment:
(aside!) Boy it would really be neat if we could use actual inheritance better, right? :) All this copy pasted code... I don't see any other way right now, but I should seriously think about trying to extract some of this stuff...) (end aside)